### ISSUE TITLE
deep-review: ai6-fl-attack — L1 correctness & L4 scale findings

### ISSUE THREAD

**Issue by 5h4d0wn1k** [5h4d0wn1k/ai6-fl-attack#4]
## deep-review: ai6-fl-attack

Findings-only audit of `fl_attack.py` (658-line single module), `tests/test_fl_attack.py` (7 tests), and repo docs. All L1–L5 claims cite real `file:line` references and were **reproduced by executing the code** with numpy 2.5.3 (all 7 existing tests pass). Scope facts: one Python module, numpy the sole dependency with **no manifest file** (`requirements.txt`/`pyproject.toml` absent), **no `.github/workflows/` CI** (verified via clone and GitHub API), `unittest` suite only.

---

### L1 Code correctness

**High — zero-variance update pool is inverted into "everyone is Byzantine": all client updates silently discarded.**
`fl_attack.py:337-338`:
```python
if std_norm == 0:
    return list(range(len(norms))), norms
```
When all clients' update norms are identical (identical duplicate uploads — the *no-attack* case), `detect_outliers` returns **every** index as an outlier. The correct semantic is the opposite: zero variance ⇒ zero outliers ⇒ return `[]`. Verified by execution: two identical updates return `[0, 1]`, and `filter_byzantine` (`fl_attack.py:345-355`) then returns an **empty client pool** `[]`, dropping every update. The demo only survives because `run_experiment` guards at `fl_attack.py:493` (`if len(outlier_ids) > 0 and len(outlier_ids) < len(updates_w)`), which is why the test suite passes. Any direct consumer of `detect_outliers`/`filter_byzantine` loses the entire round. Fix: return `[]` when `std_norm == 0`; add the regression test in L6.

**High — `GradientInverter` does not actually invert gradients; the "recovered" sample is an unbounded random-init fixed point.**
`fl_attack.py:246-276`: the shared update is copied into `wfake`/`bfake` and then used **as the weights of a fresh network** for forward/backward on a random `X_recovered` (`fl_attack.py:234`). A real gradient inversion (DLG / Inverting Gradients) must solve `argmin_X || ∇θ L(X,y) − Δθ_shared ||` — i.e. compare the *gradient produced by the recovered sample* with the transmitted delta. Here the transmitted update is never matched against any gradient; nothing in the loop references the update as an optimization target. Verified: the true client sample has ‖X‖≈3.82, while `invert()` returns ‖X‖≈0.036 (mean≈0.001, std≈0.013) after 300 iterations at lr 0.05 — shaped only by the random init scale (`fl_attack.py:234`) and the norm-clip at `fl_attack.py:274`, orders of magnitude below any plausible reconstruction. `batch_invert` (`fl_attack.py:278-283`) is likewise not a batch-objective inversion: it stacks `batch_size` **independent single-label** reconstructions. The README headline "gradient inversion — single and batch sample recovery" (README.md:35) is therefore not grounded in the code. Fix-outline: a true gradient-matching inner loop (`compute grad of a dummy net at X_recovered; step X to minimize ‖grad − Δ‖`), and report a reconstruction metric (cosine similarity vs. the true sample).

**Med — Krum silently no-ops when `Byzantine_count ≥ n−1`.**
`fl_attack.py:152-156`: `score = np.sum(sorted_dists[1:n - f])`. When `n − f ≤ 1` the slice is empty, every score is 0, and `np.argmin` returns index 0 arbitrarily. Verified: with `n=4`, `Byzantine_count=3` **and** `4`, `aggregate_krum` returns client 0's update without any error. Validate `f < n − 1` and raise a clear `ValueError`, or clamp; add a regression test.

**Med — `label_flipping_poison` does not perform label flipping, and is dead code.**
`fl_attack.py:215-221` negates the model-update values and `np.roll`s the bias columns along axis 1. Neither operation flips **data labels** — "label flipping" in FL means relabeling the client's training set before local training. The method is never called from `run_experiment` (which uses `sign_flip_poison` at `fl_attack.py:413`/`524` and `scale_poison` at `fl_attack.py:487`) and has no test. Either implement a real label-flip path (relabel `self.y` before `local_train`, `fl_attack.py:64`) or delete the method.

**Med — `aggregate_mean` accepts non-normalized weights and silently mis-scales the global model.**
`fl_attack.py:121-131` normalizes only the default weights (`np.ones(n)/n`); a caller-supplied `weights` array is used verbatim. Verified: `weights=[1,1]` on two updates returns an aggregate with ~2× the norm of a mean — the global model is silently scaled. `run_experiment` never passes explicit weights, but the parameter is public API and the L8 FedAvg feature would rely on it. Assert `np.isclose(weights.sum(), 1.0)` or normalize inside the function.

**Low — README exit-code contract is wrong.**
README.md:62 says bad arguments exit 1; argparse exits **2** (verified: `--clients nope` → exit 2). Report-write failure does exit 1, but only as a side effect of an unhandled exception (see L3).

---

### L2 Security

**No substantive issues found at this layer.** The tool is fully offline and self-contained (numpy-only): no network I/O, no `subprocess`, no `eval/exec`, no HTML/XML rendering, no untrusted-input parsing, and no credential-shaped strings anywhere in the tree (read every file: `fl_attack.py:1-658`, `tests/test_fl_attack.py`, README, SECURITY, ETHICS, SCOPE, CONTRIBUTING, CODE_OF_CONDUCT, LICENSE). The CLI `--output` writes a user-specified path after `os.makedirs(exist_ok=True)` (`fl_attack.py:647-651`) — a normal CLI contract, no traversal surface beyond the operator's own filesystem. Governance docs are appropriately strong for an attack-tool repo (README.md:1-5, ETHICS.md:8-27, SCOPE.md:7-25, SECURITY.md:3-34). The one hygiene gap is dependency provenance: `numpy` is unpinned and there is no lockfile/SBOM (see **L7**).

---

### L3 Error handling & resilience

**Med — zero error handling or logging anywhere; every failure is a raw traceback.**
There is no `try/except` and no `logger` in `fl_attack.py`. Verified: `--output /nonexistent_dir/x.json` raises an uncaught `PermissionError` traceback at `fl_attack.py:649` (exit 1); plain `open()` at `fl_attack.py:650-651` has no handling for `PermissionError`/`IsADirectoryError`. CLI boundary *values* are unvalidated — `--clients 0`, `--rounds 0`, `--features 0`, `--classes` mismatches (`fl_attack.py:630-640`) fall through to numpy-level crashes in the middle of `run_experiment` rather than a one-line message. `format_report` indexes nested result keys directly (`fl_attack.py:583-613`), so future schema drift fails with `KeyError` at print time. Fix-outline: wrap `main()` in a top-level error handler that prints one-line errors and returns a non-zero code; validate `clients/rounds/features ≥ 1` and `num_classes ≥ 2` up front.

---

### L4 Performance & scale

**Med — Krum is O(n²) in number of clients with `O(all_params)` memory duplication; demo-only at real federation size.**
`fl_attack.py:142-147` materializes an `n×n` matrix where each distance is a full flattened-update vector norm, and line 140 stacks **all** client updates into one dense array. For a real federation (hundreds–thousands of clients, ML-sized models) that is `O(n²·D)` time and `O(n·D)` memory — the practical ceiling before anything else. The mean aggregator (`fl_attack.py:127-130`) is a pure-Python double loop that would be a single stacked `np.array(...) @ weights` op. Trimmed mean (`fl_attack.py:169-179`) full-sorts each layer per round; `np.partition`-based trimming avoids the sort. Not a blocker for the demo (default n=6), but a bounded distance-chunking scheme plus a complexity note would make `--clients 100+` usable.

---

### L5 Data & correctness

**Med — the synthetic benchmark task has no learnable signal: X and y are independent, so every headline number is chance-level.**
Clients draw `X = rng.randn(data, input_size)` and `y = rng.randint(0, C)` independently (`fl_attack.py:21-22`); the test set is drawn from unrelated random distributions (`fl_attack.py:377-378`, y from a *different* RNG at line 378). No model can learn a signal, so baseline accuracy sits at plurality-chance and the report's `accuracy_drop` (`fl_attack.py:547`, `accuracy` at `fl_attack.py:358-360`) and defense `improvement` metrics (`fl_attack.py:563-564`) are sampling noise — not a measurement of attack/defense efficacy. Fix: make `y` class-conditional (per-class Gaussian mean shift) so the demo measures a real signal-degradation curve, and add a test asserting clean accuracy is meaningfully above chance.

**Low — dead state and unbounded history growth.**
`server.history` is only initialized and never written (`fl_attack.py:119`); `ByzantineDetector.scores_history` appends one z-score vector per `detect_outliers` call with no bound (`fl_attack.py:342`), so a long-running evaluator leaks memory. Wire `history` to per-round accuracies in `run_experiment` and cap `scores_history`.

---

### L6 Test coverage

Current suite: 7 tests, all passing (verified). Covered: `local_train` update shapes, mean-agg output shapes, sign-flip equality, z-score outlier on a 6-client pool, experiment structure keys, seed determinism, CLI JSON write. **Genuinely untested critical paths (all of which hid real bugs above):** `aggregate_krum` — no test at all (silently broken for `f ≥ n−1`); `aggregate_trimmed_mean` — no test; `GradientInverter.invert`/`batch_invert` — no test (the L1 flaw ships); `FreeRider.generate_fake_update`/`mimicking_update` — no test; `ModelUpdatePoisoner.label_flipping_poison`/`gaussian_noise_poison` — no test; `filter_byzantine` — no test; `aggregate_mean` with explicit weights — no test; zero-variance `detect_outliers` path; CLI boundary *values*; `format_report`.

Proposed concrete test cases:
1. `test_detect_outliers_zero_variance_returns_empty` — two identical updates → assert `outliers == []` (currently returns `[0,1]`; drives the L1 fix).
2. `test_krum_rejects_f_ge_n_minus_1` — `aggregate_krum(n=4, Byzantine_count=3)` raises `ValueError` instead of silently returning client 0.
3. `test_invert_reconstruction_quality` — after the gradient-matching rework, assert cosine similarity(recovered, true sample) ≥ 0.7 or that reconstruction loss decreases (regression lock for L1).
4. `test_trimmed_mean_robustness` — with honest updates, trimmed ≈ mean within tolerance; with one large outlier inserted, `Δ_trimmed << Δ_mean`.
5. `test_aggregate_mean_rejects_non_unit_weights` — `weights=[1,1]` is rejected/normalized, locking L1.
6. `test_freerider_update_stays_small` — `generate_fake_update(magnitude=0.0001)` norm stays below honest norms, guarding scenario [4].

---

### L7 Build/CI-CD

**There is no CI at all.** No `.github/workflows/` exists in the tree or on `master` (verified by clone and `GET /repos/.../contents/.github/workflows` → 404). Tests run only manually (`python3 -m unittest discover -s tests -v`, README.md:59). There is **no dependency manifest** — the sole dependency `numpy` is unpinned (README.md:43), so runs and tests are not reproducible across numpy versions. Nothing runs lint, type-check, coverage, dependency audit, secret scanning, or code scanning. The 7 tests gate nothing and cannot catch the L1 regressions above (the determinism test at `tests/test_fl_attack.py:94-100` passes while the reported metrics are meaningless under L5).

Concrete pipeline improvements:
1. Add `.github/workflows/ci.yml` (ubuntu-latest, `actions/setup-python@v5`, install `numpy ruff mypy coverage`, run `python -m unittest discover -s tests -v`, `ruff check fl_attack.py`, and coverage upload) so the 7 tests plus lint actually gate PRs.
2. Add `pyproject.toml` pinning `numpy>=1.26,<3`, a lockfile (`uv lock` or pip-tools), and a `requirements-dev.txt` — closes the L2 dependency-provenance gap and makes the environment reproducible.
3. Add a `dependabot.yml` (pip + github-actions), a CodeQL analysis workflow, and a `gitleaks`/secret-scan job — cheap and high-value for a security-adjacent repo.

---

### L8 Features (tied to existing architecture)

1. **Non-IID client partitions (Dirichlet skew)** — all clients currently sample from the same i.i.d. distribution (`fl_attack.py:21-22`); adding per-client class-distribution skew reflects real FL data heterogeneity and is exactly what makes poisoning harder to detect in practice.
2. **Data-size-weighted FedAvg** — `aggregate_mean` already accepts `weights` (`fl_attack.py:121-131`) but `run_experiment` never passes any; wire per-client `data_size` weights into scenario [1] (`fl_attack.py:390`) for a correct FedAvg baseline.
3. **Complete `label_flipping_poison` or remove it** — implement a real label-flip client path (relabel `self.y` then `local_train`, `fl_attack.py:64`) so scenario [2] can compare sign-flip vs. label-flip impact using the currently dead method (`fl_attack.py:215-221`).
4. **True gradient-matching inversion plus a similarity metric** — rework `GradientInverter` to the DLG objective and add a `reconstruction_similarity` field to the `gradient_inversion` result block (`fl_attack.py:549-553`); without this the published `inverted_norm` cannot be validated.
5. **Per-round history and accuracy curves** — `server.history` is dead state (`fl_attack.py:119`); record per-round global accuracy in `run_experiment` and emit it in the JSON report so the demo produces real learning curves instead of only final scalars.
6. **Defense-sweep mode** — scenarios [1]/[5]/[6] each rebuild a fresh server and poisoner (`fl_attack.py:380-534`); a `--defense-sweep` running mean, Krum, and trimmed-mean over the *same* poisoned update pool would turn the `defenses` block (`fl_attack.py:560-565`) into a directly comparable table.

---

## Severity summary

- **Critical:** none
- **High:** 2 — zero-variance Byzantine data-loss inversion (L1); non-gradient-matching "inversion" (L1).
- **Med:** 7 — Krum silent no-op for large `Byzantine_count` (L1); fake label-flip method (L1); non-normalized `aggregate_mean` weights (L1); chance-level synthetic task invalidating reported metrics (L5); zero error handling/logging (L3); O(n²) Krum at scale (L4); no CI/manifest/pins (L7).
- **Low:** exit-code doc mismatch (L1); dead `server.history` and unbounded `scores_history` (L5).

## Proposed fix-PR outline

1. Fix `detect_outliers` zero-variance → return `[]` (L1, with test 1).
2. Rework `GradientInverter` to a true gradient-matching objective and add a cosine-similarity metric to results (L1, with test 3).
3. Validate `Byzantine_count < n − 1` in `aggregate_krum`; validate/normalize `weights` in `aggregate_mean` (L1, with tests 2, 5).
4. Implement real `label_flipping_poison` semantics or delete the dead method (L1).
5. Make the synthetic task class-conditional so metrics measure a real signal (L5, with a chance-level assertion test).
6. Add `main()` argument validation and a top-level error handler returning one-line errors with non-zero exit codes (L3).
7. Add the L6 tests and the L7 CI (`ci.yml`, `pyproject.toml` + lockfile, dependabot, CodeQL, secret scan).

Reviewed from a fresh `--depth 1` clone; every L1–L5 claim reproduced by executing the module with numpy 2.5.3 (all 7 existing tests pass). Findings only — no PRs created per protocol.
